### PR TITLE
feat: add ResourceFieldSelector, support env.valueFrom.resourceFieldRef

### LINKED ISSUE #35: frontend.Env.ValueFrom.ResourceFieldRef is not supported
k8s api :

```go
// EnvVarSource represents a source for the value of an EnvVar.
type EnvVarSource struct {
	// Selects a field of the pod: supports metadata.name, metadata.namespace, `metadata.labels['<KEY>']`, `metadata.annotations['<KEY>']`,
	// spec.nodeName, spec.serviceAccountName, status.hostIP, status.podIP, status.podIPs.
	// +optional
	FieldRef *ObjectFieldSelector `json:"fieldRef,omitempty" protobuf:"bytes,1,opt,name=fieldRef"`
	// Selects a resource of the container: only resources limits and requests
	// (limits.cpu, limits.memory, limits.ephemeral-storage, requests.cpu, requests.memory and requests.ephemeral-storage) are currently supported.
	// +optional
	ResourceFieldRef *ResourceFieldSelector `json:"resourceFieldRef,omitempty" protobuf:"bytes,2,opt,name=resourceFieldRef"`
	// Selects a key of a ConfigMap.
	// +optional
	ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
	// Selects a key of a secret in the pod's namespace
	// +optional
	SecretKeyRef *SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
}
```

### PR DESCRIPTION
**Pull request by howieyuen**
close #35

### CODE REVIEW

File: base/pkg/kusion_models/kube/frontend/container/env/env.k [hunk 1]
```diff
@@ -70,17 +76,36 @@ schema ObjectFieldSelector:
 
     Attributes
     ----------
-    apiVersion: str, default is Undefined, required.
+    apiVersion: str, default is v1, optional.
```
Comment (Peefy) [inline base/pkg/kusion_models/kube/frontend/container/env/env.k:79]: default is Undefined

File: base/pkg/kusion_models/kube/frontend/container/env/env.k [hunk 2]
```diff
@@ -70,17 +76,36 @@ schema ObjectFieldSelector:
...
         A Container-level attribute.
-        Version of the schema the FieldPath is written in terms of.
+        Version of the schema the FieldPath is written in terms of, defaults to "v1".
     fieldPath: str, default is Undefined, required.
         A Container-level attribute.
         Path of the field to select of an object.
     """
 
-    apiVersion: str
+    apiVersion?: str
```
Comment (Peefy) [inline base/pkg/kusion_models/kube/frontend/container/env/env.k:87]: seems to loss the default value `"v1"`?

File: base/pkg/kusion_models/kube/frontend/container/env/env.k [hunk 3]
```diff
@@ -70,17 +76,36 @@ schema ObjectFieldSelector:
...
+    resource: str, default is Undefined, required.
+        A Container-level attribute.
+        Resource to select.
+    divisor: int | Unit, default is 1, optional.
+        A Container-level attribute.
+        Specifies the output format of the exposed resources, defaults to 1
+    """
+    containerName?: str
+    resource: str
+    divisor?: int | Unit
```
Comment (Peefy) [inline base/pkg/kusion_models/kube/frontend/container/env/env.k:107]: seems to loss the default value `1`?

**Review by Peefy — COMMENTED** [review]
Need more example cases about `ResourceFieldSelector`

**Comment by howieyuen** [comment]
> Need more example cases about `ResourceFieldSelector`

field divisor is a little special for its type is `Unit` or `int`, but `Unit` has been predefined already.
I think no more redundant explanation.

**Review by Peefy — APPROVED** [review]
LGTM

File: base/pkg/kusion_models/kube/frontend/container/env/env.k [hunk 4]
```diff
@@ -39,15 +40,20 @@ schema EnvValueFrom:
...
 
     fieldRef?: ObjectFieldSelector
     configMapKeyRef?: ObjectKeySelector
     secretKeyRef?: ObjectKeySelector
+    resourceFieldRef?: ResourceFieldSelector
 
     check:
-        notfieldRef or notconfigMapKeyRef or notsecretKeyRef, "may not have more than one field specified at a time"
-        fieldRef or configMapKeyRef or secretKeyRef, "must specify one of: `fieldRef`, `configMapKeyRef` or `secretKeyRef`"
+        not fieldRef or not configMapKeyRef or not secretKeyRef or not resourceFieldRef, "may not have more than one field specified at a time"
```
Comment (amyXia1994) [inline base/pkg/kusion_models/kube/frontend/container/env/env.k:55]: 1. The check expression seems wrong, I think you mean only one field can be set at the same time, but the expression results to True as long as one of them is not set

2. This error message is a bit obscure, maybe it's better to use "the env value can not come from more than one source"

Comment (howieyuen) [reply base/pkg/kusion_models/kube/frontend/container/env/env.k]: nice catch, 填坑不找挖坑人，反叫踩坑人…… 😭

File: base/pkg/kusion_models/kube/frontend/container/env/env.k [hunk 5]
```diff
@@ -39,15 +40,22 @@ schema EnvValueFrom:
...
     secretKeyRef?: ObjectKeySelector
+    resourceFieldRef?: ResourceFieldSelector
 
     check:
-        notfieldRef or notconfigMapKeyRef or notsecretKeyRef, "may not have more than one field specified at a time"
-        fieldRef or configMapKeyRef or secretKeyRef, "must specify one of: `fieldRef`, `configMapKeyRef` or `secretKeyRef`"
+        fieldRef and not configMapKeyRef and not secretKeyRef and not resourceFieldRef
+        not fieldRef and configMapKeyRef and not secretKeyRef and not resourceFieldRef
+        not fieldRef and not configMapKeyRef and secretKeyRef and not resourceFieldRef
+        not fieldRef and not configMapKeyRef and not secretKeyRef and resourceFieldRef, "must specify one of: `fieldRef`, `configMapKeyRef`, `secretKeyRef` or `resourceFieldRef`"
```
Comment (amyXia1994) [inline base/pkg/kusion_models/kube/frontend/container/env/env.k:58]: the error message need to apply to all the check expressions

Comment (howieyuen) [reply base/pkg/kusion_models/kube/frontend/container/env/env.k]: got it

**Review by amyXia1994 — APPROVED** [review]
lgtm

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**